ADR/0023-release-and-releaseplan-authorization.md

```diff
@@ -0,0 +1,55 @@
+# Release and ReleasePlan Authorization
+
+Date: 2023-04-17
+
+## Status
+
+Accepted
+
+## Context
+
+* It is imperative to know which associate has authorized a change to production in order to have accountability for changes.
+* Flexibility is required when specifying authorization either on a per-release basis or as a standing authorization.
+
+## Decision
+
+### Architecture Overview
+#### Terminology
+
+- `Release`: A CRD that describes what should be released and using which plan.
+- `ReleasePlan`: A CRD that describes where and how an application should be released.
+- `ReleaseStrategy`: A CRD that describes which pipeline to execute along with specific pipeline parameters.
+
+#### Architecture
+
+![](../diagrams/ADR-0023/ReleaseAndReleasePlanAuthorization.jpg)
+
+##### Capturing authorization data
+* The idea is that when a release pipeline executes, the `Pipeline`'s `Tasks` have access to the name of the associate that authorized the change in production.
+* `Authorization` for a particular `Release` should be obtained by capturing the user that has created the object.
+  * The `status` section of the `Release` CR should be updated with to specify the `author`.
+* A `Standing Authorization` for an application should be obtained by capturing the user that creates or updates the `ReleasePlan` CR **when** `annotations.standing-authorization` is set to `true`
+  * Since we expect ReleasePlans to be updated by users who may not intend to provide a `Standing Authorization` (i.e. Change `ReleaseStrategy`), it is important to intercept only when the above annotation is set.
```

Comment:
What would happen if a standing authorization was configured for one individual and then another individual changed it. Would the standing authorization be changed to the new individual?

Instead of using standing-authorization as a boolean, I think it might make more sense for it to be a list of users. Therefore, if any one of those users modifies the `ReleasePlan` then their authorization will be given. If any other user modifies the ReleasePlan, then authorization will need to be revoked. 

I don't know the details of the release process, but should we actually block all releases if a ReleasePlan is modified by an individual that doesn't have standing authorization until authorization has been given? This would be to prevent the ReleasePlan from being unintentionally (or intentionally) changed to an undesired state and not detecting the unintended state until too late.

Reply:
> What would happen if a standing authorization was configured for one individual and then another individual changed it. Would the standing authorization be changed to the new individual?

Yes, it would be changed.

> Instead of using standing-authorization as a boolean, I think it might make more sense for it to be a list of users. Therefore, if any one of those users modifies the ReleasePlan then their authorization will be given. If any other user modifies the ReleasePlan, then authorization will need to be revoked.

First, a change to the standing-authorization must be done by changing a specific annotation. By doing so, we capture that user and record it. That becomes the new  standing-authorization.

> I don't know the details of the release process, but should we actually block all releases if a ReleasePlan is modified by an individual that doesn't have standing authorization until authorization has been given? This would be to prevent the ReleasePlan from being unintentionally (or intentionally) changed to an undesired state and not detecting the unintended state until too late.

Yes, the ADR mentions that we will mark the Release as failed (`ValidationError`) if we do not have a valid author (either from the Release or the ReleasePlan CRs). Because of this, no pipeline would even be triggered.

Therefore, with no authorization (singular or standing), releases are blocked.

Reply:
I will rephrase my concern with a potential solution. Should anyone with a role of `contributor` or higher be "trusted" enough to sign off on a release? 

If we might consider that anyone with these roles would not be trusted, then we shouldn't enable the `standing-authorization` to be updated to just anyone and we need to ensure that there are additional checks in place. The potential checks as I mentioned them would have no effect, however, because if someone can modify a release then they can modify a releaseplan. Therefore, the behavior of `standing-authorization` as mentioned should be fine. If any checks are needed, there should be additional parameters on the releaseplanadmission. This falls into the other conversation on this ADR and does not need to be re-hashed here.

I think I have clarified my concerns, so feel free to resolve this conversation if you think we are on the same page.

Reply:
we are on the same page.